datafusion/physical-expr/src/intervals/interval_aritmetic.rs

```diff
@@ -508,6 +580,83 @@ fn cast_scalar_value(
     ScalarValue::try_from_array(&cast_array, 0)
 }
 
+/// This function calculates the final cardinality result by inspecting the endpoints of the interval.
+fn calculate_cardinality_based_on_bounds(
+    lower_open: bool,
+    upper_open: bool,
+    diff: u64,
+) -> u64 {
+    match (lower_open, upper_open) {
+        (false, false) => diff + 1,
+        (true, true) => diff - 1,
+        _ => diff,
+    }
+}
+
+trait OneTrait: Sized + std::ops::Add + std::ops::Sub {
+    fn one() -> Self;
+}
+
+macro_rules! impl_OneTrait{
+    ($($m:ty),*) => {$( impl OneTrait for $m  { fn one() -> Self { 1 as $m } })*}
+}
+impl_OneTrait! {u8, u16, u32, u64, i8, i16, i32, i64, f32, f64}
+
+/// This function either increments or decrements its argument, depending on the `DIR` value. If `true`, it increments; otherwise it decrements the argument.
+fn increment_decrement<const DIR: bool, T: OneTrait + SubAssign + AddAssign>(
+    mut val: T,
+) -> T {
+    if DIR {
+        val.add_assign(T::one());
+    } else {
+        val.sub_assign(T::one());
+    }
+    val
+}
+
+/// This function returns the next/previous value depending on the `DIR` value.
+/// If `true`, it returns the next value; otherwise it returns the previous value.
+fn get_next_value<const DIR: bool>(value: ScalarValue) -> ScalarValue {
```

Comment:
Perhaps this would be more consistent if it were added to ScalarValue itself

```rust
impl ScalarValue  {
  fn next_value(&self) -> ScalarValue {
..
}
```